File: solutions/zig/food-chain/1/food_chain.zig
const std = @import("std");

const Animal = enum { fly, spider, bird, cat, dog, goat, cow, horse };

const Verse = struct {
    animal: Animal,
    that: []const u8,
};

const verses: [8]Verse = .{
    .{ .animal = .horse, .that = "She's dead, of course!" },
    .{ .animal = .cow, .that = "I don't know how she swallowed a cow!\n" },
    .{ .animal = .goat, .that = "Just opened her throat and swallowed a goat!\n" },
    .{ .animal = .dog, .that = "What a hog, to swallow a dog!\n" },
    .{ .animal = .cat, .that = "Imagine that, to swallow a cat!\n" },
    .{ .animal = .bird, .that = "How absurd to swallow a bird!\n" },
    .{ .animal = .spider, .that = "It wriggled and jiggled and tickled inside her.\n" },
    .{ .animal = .fly, .that = "" },
};

inline fn spiderCompl(animal: Animal) []const u8 {
    return switch (animal) {
        .spider => " that wriggled and jiggled and tickled inside her",
        else => "",
    };
}

inline fn flyCompl(animal: Animal) []const u8 {
    return switch (animal) {
        .fly => "I don't know why she swallowed the fly. Perhaps she'll die.",
        else => "",
    };
}

pub fn recite(buffer: []u8, start_verse: u32, end_verse: u32) []const u8 {
    var s = std.io.fixedBufferStream(buffer);
    for (start_verse - 1..end_verse) |i| {
        if (s.pos > 0) s.writer().writeAll("\n") catch unreachable;

        // intro
        const swallowed = verses[verses.len - i - 1];
        s.writer().print("I know an old lady who swallowed a {s}.\n", .{@tagName(swallowed.animal)}) catch unreachable;
        s.writer().writeAll(swallowed.that) catch unreachable;
        s.writer().writeAll(flyCompl(swallowed.animal)) catch unreachable;
        if (i == 7) break;

        // swalloed
        for (verses.len - i..verses.len) |j| {
            const from = verses[j - 1];
            const to = verses[j];
            s.writer().print("She swallowed the {s} to catch the {s}{s}.\n", .{ @tagName(from.animal), @tagName(to.animal), spiderCompl(to.animal) }) catch unreachable;
            s.writer().writeAll(flyCompl(to.animal)) catch unreachable;
        }
        if (i < end_verse - 1) s.writer().writeAll("\n") catch unreachable;
    }
    return s.getWritten();
}

File: solutions/lua/food-chain/1/food-chain.lua
local animals = { "fly", "spider", "bird", "cat", "dog", "goat", "cow", "horse" }
local function verse(which)
  local intro_ = { ""
  , "It wriggled and jiggled and tickled inside her."
  , "How absurd to swallow a bird!"
  , "Imagine that, to swallow a cat!"
  , "What a hog, to swallow a dog!"
  , "Just opened her throat and swallowed a goat!"
  , "I don't know how she swallowed a cow!"
  , "She's dead, of course!" }
  local intro = ("I know an old lady who swallowed a %s.\n%s"):format(animals[which], intro_[which])
  if which == #animals then return intro .. "\n" end

  local function swallowed()
    local descr = { [2] = " that wriggled and jiggled and tickled inside her" }
    local result = setmetatable({}, { __index = table })
    for i = which - 1, 1, -1 do
      result:insert(("She swallowed the %s to catch the %s%s."):format(animals[i + 1], animals[i], descr[i] or ""))
    end
    return #result == 0 and "" or "\n" .. result:concat("\n") .. "\n"
  end
  return intro .. swallowed() .. "I don't know why she swallowed the fly. Perhaps she'll die.\n"
end

local function verses(from, to)
  local result = setmetatable({}, { __index = table })
  for i = from, to do result:insert(verse(i)) end
  return result:concat("\n") .. "\n"
end

local function sing() return verses(1, #animals) end

return { verse = verse, verses = verses, sing = sing }
